database/skylink.go

```diff
@@ -0,0 +1,59 @@
+package database
+
+import (
+	"time"
+
+	accdb "github.com/SkynetLabs/skynet-accounts/database"
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.sia.tech/siad/crypto"
+)
+
+var (
+	// ErrInvalidSkylink is the error returned when the passed skylink is
+	// invalid.
+	ErrInvalidSkylink = "invalid skylink"
+
+	// SkylinkStatusNew is the status of the skylink when it's created.
+	SkylinkStatusNew = "new"
+	// SkylinkStatusScanning is the status of the skylink while it's being
+	// scanned.
+	SkylinkStatusScanning = "scanning"
+	// SkylinkStatusComplete is the status of the skylink after it's scanned.
+	SkylinkStatusComplete = "complete"
+)
+
+// Skylink represents a skylink in the queue and holds its scanning status.
+type Skylink struct {
+	ID                   primitive.ObjectID `bson:"_id,omitempty" json:"-"`
+	Hash                 crypto.Hash        `bson:"hash" json:"hash"`
+	Skylink              string             `bson:"skylink" json:"skylink"`
+	Status               string             `bson:"status" json:"status"`
+	Infected             bool               `bson:"infected" json:"infected"`
+	InfectionDescription string             `bson:"infection_description" json:"infection_description"`
+	// Timestamp marks the last status change that happened to the record. It
+	// can be the time when it was created, locked for scanning, or scanned.
+	Timestamp time.Time `bson:"timestamp" json:"timestamp"`
+}
+
+// LoadString parses a skylink from string and populates all required fields.
+func (s *Skylink) LoadString(skylink string) error {
+	if !accdb.ValidSkylinkHash(skylink) {
+		return errors.New(ErrInvalidSkylink)
+	}
+	s.Skylink = skylink
+	var sl skymodules.Skylink
+	err := sl.LoadString(skylink)
+	if err != nil {
+		return errors.AddContext(err, ErrInvalidSkylink)
+	}
+	s.Hash = crypto.HashObject(sl.MerkleRoot())
```

Comment:
1. This is how these hashes are being created in `managedBlocklistHash` in `skyd`. Now I see that I'm missing the proper handling of v2 skylinks, which makes me wonder where I got this code from. It might be from skynet-js - I will  check whether we need that handling there, as well.
2. This is not great. I read the code `MerkleRoot()` now and it is evident from it that it should not be used for v1 skylinks but important information like that should be mentioned in the docstring of the method as well. The issue of the critical itself will be resolve when I mimic the behaviour of `managedBlocklistHash`.
3. That's not great. We need to figure out how to work with that while also getting rid of the skylinks. Or revisit the requirement to not store skylinks.

Reply:
In skyd we assume that if you block a skylink pointing to a base sector you also want all other skylinks pointing to that base sector to be blocked.

So all you need to do is fetch the data using the skylink and verify it. Then there are 2 scenarios.
1. it contains malware -> block link (all other links pointing to that sector will be blocked implicitly)
2. it doesn't contain malware -> it's clean, other links to that sector might contain malware but there is no way to verify if such data exists

For v2 links just block the v2 link. That should block the underlying v1 link. We might need to add special handling for v2 links later since they can be updated without us knowing but that's probably out of scope for the initial MR. In the worst case we need to fully block v2 links and not just the underlying v1 links in skyd.